File: sorting/DutchNationalFlag.java
package sorting;

/*
 *Explanation:
The code above solves the Dutch National Flag problem. It takes an array of integers and a pivot element as input, and sorts the array in the Dutch National Flag order. The Dutch National Flag problem involves partitioning the array into three sections based on the pivot element.

The dutchNationalFlagSort method uses three pointers: low, mid, and high. The low pointer represents the boundary for elements less than the pivot, the mid pointer represents the boundary for elements equal to the pivot, and the high pointer represents the boundary for elements greater than the pivot.

The method uses a while loop that continues until the mid pointer surpasses the high pointer. Inside the loop, it compares the element at the mid index with the pivot. If the element is less than the pivot, it swaps it with the element at the low index and increments both low and mid. If the element is greater than the pivot, it swaps it with the element at the high index and decrements high. If the element is equal to the pivot, it increments mid.

The swap method is a helper function that swaps two elements in the array.

In the main method, an example array and pivot value are provided. The dutchNationalFlagSort method is called with these values, and the sorted array is printed.

Time Complexity: The time complexity of the Dutch National Flag algorithm is O(n), where n is the length of the input array.

Space Complexity: The space complexity is O(1) as the algorithm sorts the array in-place without using any additional data structures. 
**/
import java.util.Arrays;

public class DutchNationalFlag {

    /**
     * Sorts an array of integers in the Dutch National Flag order.
     * The Dutch National Flag problem partitions the array into three sections:
     * - All elements less than the pivot are placed before it.
     * - All elements greater than the pivot are placed after it.
     * - All elements equal to the pivot are placed in the middle.
     *
     * @param array the array of integers to be sorted
     * @param pivot the pivot element
     */
    public static void main(String[] args) {
        int[] array = {2, 2, 1, 1, 0, 0, 2, 1, 0};
        int pivot = 1;

        dutchNationalFlagSort(array, pivot);

        System.out.println("Sorted Array: " + Arrays.toString(array));
    }
    
    public static void dutchNationalFlagSort(int[] array, int pivot) {
        int low = 0; // Pointer for elements less than the pivot
        int mid = 0; // Pointer for elements equal to the pivot
        int high = array.length - 1; // Pointer for elements greater than the pivot

        while (mid <= high) {
            if (array[mid] < pivot) { // Current element is less than the pivot
                swap(array, low, mid); // Swap current element with element at the low index
                low++; // Increment low pointer
                mid++; // Increment mid pointer
            } else if (array[mid] > pivot) { // Current element is greater than the pivot
                swap(array, mid, high); // Swap current element with element at the high index
                high--; // Decrement high pointer
            } else { // Current element is equal to the pivot
                mid++; // Increment mid pointer
            }
        }
    }

    /**
     * Swaps two elements in an array.
     *
     * @param array the array
     * @param i     the index of the first element
     * @param j     the index of the second element
     */
    private static void swap(int[] array, int i, int j) {
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }
}

File: Arrays/dutch_national_flag.cpp
/*
    Given an array nums with n objects colored red, white, or blue, sort them in-place so that objects of the same color are adjacent, with the colors in the order red, white, and blue.

    We will use the integers 0, 1, and 2 to represent the color red, white, and blue, respectively.

    Input: nums = [2,0,2,1,1,0]
    Output: [0,0,1,1,2,2]

    Explanation:


    The algorithm partitions the array into three sections: elements with the value 0, elements with the value 1, and elements 
    with the value 2. It uses three pointers, `start`, `low`, and `end`, to keep track of the boundaries between these sections.

    The algorithm iterates through the array using the `low` pointer. Here's how it works:

    1. If the element at `low` is 0, it means it should be in the first section. In this case, the algorithm swaps the element 
       with the element at the `start` position, increments both `start` and `low` pointers, and moves the `low` pointer to the 
       next element to process.

    2. If the element at `low` is 1, it means it should be in the second section. In this case, the algorithm simply increments 
       the `low` pointer and moves to the next element to process.

    3. If the element at `low` is 2, it means it should be in the third section. In this case, the algorithm swaps the element 
       with the element at the `end` position, decrements the `end` pointer, and moves the `low` pointer to the next element to process. The reason for moving the `low` pointer to the next element is to recheck the value after the swap, as the swapped element could be 0 or 1.

    The algorithm continues these steps until the `low` pointer surpasses the `end` pointer, indicating that all elements have 
    been processed.

    The "Dutch National Flag" algorithm has a time complexity of O(n), where n is the size of the input array. 
    It performs a single pass through the array, ensuring that all elements are correctly placed in their respective sections.

    Note that the code assumes the input vector `nums` contains only values 0, 1, and 2, and it modifies the vector in-place 
    to achieve the sorted order.

    Time complexity: O(n)
    Space complexity: O(1)
*/



#include <bits/stdc++.h>
using namespace std;

void sortColors(vector<int> &nums)
{
    // created 3 variables start , low and end which are pointing start and low which are pointing to first index , end is pointing to last index .

    int start = 0, low = 0, end = nums.size() - 1;
    while (low <= end)
    {
        if (nums[low] == 0) // checking if element of low is 0 . If yes then  swap to start and low .
        {
            swap(nums[low], nums[start]);
            start++, low++;
        }
        else if (nums[low] == 1) // checking if element at low index is 1 , If yes then increase the index by 1 .
        {
            low++;
        }
        else // else swap the element of low index to end .
        {
            swap(nums[low], nums[end]);
            end--;
        }
    }
}
int main()
{
    vector<int> nums{2, 0, 2, 1, 1, 0};
    sortColors(nums);

    // Printing array's elements ..
    for (auto i : nums)
    {
        cout << i << " ";
    }
}
